Improve CircleCI Workflow for Multi-Arch

dpipemazo opened the pull request:
Atom Orb has been restructured to do the following:

- Large overhaul/restructure of the orb for better integration with new
CircleCI features such as pipelines and matrix configurations.
- Atom orb now should mainly be used with the following commands: `atom/build_buildx`,
`atom/test`, `atom/deploy`, and `atom/deploy_latest`. This should suit the needs of most users. There
are sub-commands broken out in the orb if you need something slightly
different/special for your build.
- The only CircleCI vairables now needed are `DOCKERHUB_USER` and `DOCKERHUB_PASSWORD`.
All other variables have been moved to pipeline parameters in the `.circleci`
file to make it a bit easier to configure. Other build variables should
be removed from your project when possible.
- Atom orb now only depends on the `DOCKERHUB_USER` and `DOCKERHUB_PASSWORD`
hidden variables. Other than that everything is broken out to make it
more what's happening more clear to the user.
- Atom orb supports/recommends using matrix configuration for building
for multi-arch. All jobs support the following fields: `variant` and `platform`.
A `variant` is either "stock" (i.e. vanilla), "opengl", "cuda", etc. A `platform`
is currently either `amd64` or `aarch64`.
- `atom/deploy` now supports the `target_tag_cmd` field which allows you to
run a command-line text replacement (grep, sed, etc.) on the tag field. By
default, this field removes `-stock` and `-amd64` from the auto-generated tag
s.t. stock builds for intel machines produce tags such as "development-X",
"master-X", etc. as before. The auto-generated tag for builds automatically
appends the variant and platform to the build product. In general you shouldn't
need to mess with this field as it's mainly a crutch to let us use the matrix
features and still get all of our tags as expected but you might want to
use it at some point.

In accordance with this, the CircleCI config and examples/templates in this repo have been updated. Several elements have been ported over to using the new orb to test the reusability/portability of the orb and I think it's in decent shape.

One thing to note in this PR is the concept of `variant` and `platform` when it comes to building our images.

A `variant` is currently one of:
- stock
- opengl
- opengl-cuda
- cuda

A `platform` is currently one of
- amd64
- aarch64

It's important to fully break these out as in the future we're going to want to be able to support openGL builds for aarch64, another platform that's perhaps optimized for jetson, etc.

When building with the orb, we now add `-<variant>-<platform>` to the end of every tag that we previously had, i.e. `development-1234` would now also be tagged as `development-1234-stock-amd64`. We're also pushing additional tags to ensure that there is also still a `development-1234` tag pushed and in general all tags are now pushed twice: once with the full `-<variant>-<platform>` suffix and once where we strip off `-stock` and `-amd64` so that we get the same tags that we're used to. This same concept applies to all elements using the orb. Special care has also been taken to ensure that `latest` still works as does `aarch64` mapping to the aarch64-equivalent of `latest`. 

With this switch we're also embracing CircleCI's new pipelines and will be using the pipeline number instead of the CircleCI build job number to tag builds. This is easier in their new UI.

aywaldron commented on .circleci/config.yml:
```diff
@@ -152,72 +226,85 @@ commands:
+          container: test_atom
           test_valgrind: << parameters.test_valgrind >>
 
   build_atom_variant:
```
Any chance this could be reused in `build_atom_with_nucleus_variant`?

dpipemazo replied:
yes, could probably simplify this a bit. Will take a look.

dpipemazo replied:
Reused.

aywaldron commented on template/.circleci/config.yml:
```diff
@@ -1,40 +1,137 @@
+    default: v1.3.0
+  dockerhub_repo:
+    type: string
+    default: elementaryrobotics/example-element
```
Didn't realize we had this template! Is it referenced in our circle docs? Maybe we don't need the example in the README if we have this instead.

dpipemazo replied:
Ah, not sure if it's referenced from the Circle docs but it's the one we point users to in the atom docs. Will remove the example in the circle docs and reference back to this.

dpipemazo commented on .circleci/atom.yml:
```diff
@@ -361,58 +409,357 @@ jobs:
+          build_args: << parameters.build_args >>
+          no_output_timeout: << parameters.no_output_timeout >>
+
+  build_large:
```
It us unfortunately necessary :( . These are jobs so that we can just do our config in workflows. You can't change a job executor from the workflow which is silly. I tried doing it through the `resource_class` and the `executor` commands from the workflow and no dice.

dpipemazo replied:
Aliases cleaned this up fairly nicely though :+1:

dpipemazo commented on .circleci/atom.yml:
```diff
@@ -70,86 +92,96 @@ commands:
-          command: docker build -f << parameters.file >> -t << parameters.image_name >>:<< parameters.image_tag >> --target=<< parameters.stage >> --build-arg << parameters.build_args >> --pull .
+          name: Build Dockerfile << parameters.working_directory >>/<< parameters.file >> stage << parameters.stage >> and tag as << parameters.image_name >>:<< parameters.image_tag >>
+          working_directory: << parameters.working_directory >>
+          command: docker build -f << parameters.file >> -t << parameters.image_name >>:<< parameters.image_tag >> << parameters.build_args >> --pull .
```
ah, nice catch. I didn't test this much and it's mainly here for legacy purposes. I would encourage everyone to use the `build_dockerfile_buildx` in the future though I will update this to get it working.

dpipemazo replied:
Reworked with aliases to be using the same partial build command.

dpipemazo commented on .circleci/atom.yml:
```diff
@@ -163,161 +195,178 @@ commands:
       - run:
-          name: Docker Tag
-          command: docker tag << parameters.source_image >> << parameters.target_image >>:<< parameters.target_tag >>
+          name: Tag image source:<< parameters.source_image >>:<< parameters.source_tag >> target:<< parameters.target_image >>:<< parameters.target_tag >> tag_cmd:<< parameters.target_tag_cmd >>
```
Yeah, I found it a bit generic to just see "push image" 3x times and not know what was being pushed -- I figured being able to see the info from the summary line would help to see which tags are generated by a certain build. Could definitely remove it and don't feel too strongly. Will check the `DOCKERHUB` env vars to see if I can find the issue there.

aywaldron replied:
Agree "push image" isn't descriptive enough, maybe something simpler like this would be easier to read:
```
Tag and push image << parameters.target_image >>:<< parameters.target_tag >> 
```

dpipemazo replied:
Sure, that would be better. Without the tag command though it’s unclear what actual tag is generated, though you can get all of the info you need from your suggestion

aywaldron replied:
Right, you could evaluate the new tag either beforehand or in there perhaps, or include the command in the name, I don't feel strongly either way.

dpipemazo replied:
Evaluating the tag is actually difficult -- CircleCI doesn't let you store temporary variables anywhere that I could find. Will clean up the command names.

dpipemazo commented on .circleci/atom.yml:
```diff
@@ -163,161 +195,178 @@ commands:
+      - checkout
+      - update_submodules
+      - docker_login
+      - build_dockerfile_buildx:
```
I don't think so -- there are a bunch more parameters enabled by `buildx`. We should all be moving to `buildx`, however it's unstable enough (a few random HTTP 400s when uploading cache) and a few issues at startup sometimes that I didn't want to force it on everyone. `build` should be deprecated with time.

dpipemazo replied:
Worked this in with aliases -- we now have basic shared build params and then additional buildx-enabled ones. They share the same base build command with additional buildx-based args.

dpipemazo commented on .circleci/atom.yml:
```diff
@@ -361,58 +409,357 @@ jobs:
+      cache_tag:
+        type: string
+      no_output_timeout:
+        <<: *default_no_output_timeout
```
ah, an alias for the parameter set is a super cool idea!

dpipemazo commented on .circleci/atom.yml:
```diff
@@ -439,14 +786,14 @@ examples:
                 file: docker-compose.yml
                 service: ${DOCKER_COMPOSE_SERVICE_NAME}
-                image_tag: ${DOCKERHUB_REPO}-${CIRCLE_WORKFLOW_ID}
+                image_tag: ${DOCKERHUB_REPO}-${CIRCLE_WORKFLOW_WORKSPACE_ID}
```
ah -- we're actually using `<< pipeline.number >>` instead of this -- I'll clean it up. Yes, it will overwrite existing image tags but I think it doesn't matter. DockerHub deletes them eventually anyway when they go unused for a while and nobody should be depending on `development-X` or `master-X` builds right now.

dpipemazo commented on .circleci/config.yml:
```diff
@@ -152,72 +226,85 @@ commands:
           cache_repo: << parameters.cache_repo >>
-          cache_tag: atom<< parameters.tag >>
-      - atom/build_stage_buildx:
+          cache_tag: cache-atom
```
The cache manifest is being pushed to `elementaryrobotics/atom` with a tag `cache-atom-<< variant >>-<< platform >>`

aywaldron replied:
Does that happen automatically as part of the `buildx` command?

dpipemazo replied:
Yes, this is what is letting us not use CircleCI DLC. It’s not required with buildx but the command I wrote uses —cache-from and —cache-to.

aywaldron replied:
Ok cool, didn't realize the push happened as part of that at the same time.

aywaldron commented on .circleci/config.yml:
```diff
@@ -600,33 +538,15 @@ workflows:
               variant: [ opengl ]
               platform: [ amd64 ]
+          stage: graphics
+          build_args: --build-arg TEST_FROM=graphics
```
Are there tests that require the graphics libs?

dpipemazo replied:
there are not, but could be. We also could break the atom dependencies/build in the graphics stage and aren't technically testing the production code which is non-ideal.

aywaldron commented on .circleci/config.yml:
```diff
@@ -943,20 +843,22 @@ workflows:
               platform: [ amd64 ]
-          base_image: debian:buster-slim
+          base_repo: debian
+          base_tag: buster-slim
```
:+1:

aywaldron commented:
All the new aliases and configs look great!

dpipemazo commented:
A few more things fell out while cleaning this up:
- I realized that while testing the graphics-based variants of atom we weren't actually testing the production image. Instead we were testing the stage pre-VNC. This opened us up to issues that the VNC could introduce and generally wasn't right.
- To handle the above, I had tried making the stage that `test` is built from parameterizable between `atom` and `graphics`. Docker doesn't seem to support this as when specifying this stage it assumes it's a dockerhub repo. It also wasn't the most elegant solution
- The right solution to the above was to revamp the build system in order to allow for chaining of Dockerfiles that add in features, such as OpenGL or a VNC. This allowed us to remove the `INSTALL_OPENGL` flag on the atom build which was wonky to begin with. Now, when we want to add a set of things to an atom base, we define it as its own Dockerfile.
- Updated the build system to make it easy to chain these Dockerfiles that add things.
- Built an `opengl-cuda` base that chains the VNC and OpenGL as the best example of this process
- Updated docs to reflect new base build process

Also, we're still seeing a few too many intermittent failures on pushing the cache from the buildx build to Dockerhub for my liking. This is particularly painful as it causes us to lose a whole build. As such, removed the `--cache-to` from the original `buildx` command and added a second, identical `buildx` command immediately following the original one what also has a suffix of `|| exit 1`. This allows us to complete our first build successfully and then optionally fail on pushing the cache without losing the whole job. I've [filed an issue on buildx](https://github.com/docker/buildx/issues/271) and hopefully we can remove it when it becomes a bit more stable. We still see occasional failures booting buildx but there's not much I can think of now that would help us with those. After this change the builds are succeeding much more frequently though we are still occasionally failing to push the cache. The main downside of that is the next build will perhaps take slightly longer if anything useful was done.

aywaldron commented on .circleci/README.md:
```diff
@@ -19,14 +19,37 @@ Upon pushing to github, a build will be started on CircleCI. This will build
+| `master` | `<< variant >>-<< platform >>` |
+| Non-`master` | `development-<< pipeline.number >>-<< variant >>-<< platform >>` |
+| tag, i.e. v1.3.1 | `v1.3.1-<< variant >>-<< platform >>` |
+
```
Should we mention the build tags that get pushed as well?

dpipemazo replied:
yeah, likely a good idea to mention that.

aywaldron commented on .circleci/config.yml:
```diff
@@ -363,39 +370,175 @@ commands:
+          input_tag: base-build
+
+      # Add in OpenGL
+      - build_atom_base_add_opengl:
```
I'm a little confused here how we have the +vnc and +opengl options, then we also have an opengl "variant". Is that redundant or they're two different things?

dpipemazo replied:
Yeah, it's confusing but they are different. The bases are used to make variants. The stock variant uses the "vanilla" base. I tried to use different terminology here to not get them confused but perhaps I would up making it more confusing in the process.

aywaldron replied:
So the opengl variant is just built using the opengl base?

dpipemazo replied:
Yeah, which is based off debian:buster-slim + the OpenGL dockerfile. Then, the `opengl-cuda` variant is built off of the opengl-cuda base which is based off of `nvidia/cuda` + the OpenGL Dockerfile.

dpipemazo commented on .circleci/atom.yml:
```diff
@@ -408,6 +406,9 @@ commands:
       build_command:
         type: string
         default: *build_basic_command
+      build_context:
```
since the build command is a YAML alias I don't think it can be appended to/edited easily.

dpipemazo replied:
how we're using it is more of a hack for now, but being able to specify the context is likely a good thing long-term.

dpipemazo commented on .circleci/atom.yml:
```diff
@@ -618,8 +638,284 @@ jobs:
+                image_tag: build-<< pipeline.number >>
+                cache_repo: << pipeline.parameters.dockerhub_repo >>
+                cache_tag: cache
+                build_args: --build-arg ATOM_IMAGE=<< pipeline.parameters.atom_repo >>:<< pipeline.parameters.atom_version >>-<< pipeline.parameters.atom_variant >>-<< pipeline.parameters.atom_platform >>
```
I don't think so because you need it before each build arg. I'm not sure what happens when passing `--build-arg` and then nothing after; I think this is more clear.

dpipemazo commented on .circleci/config.yml:
```diff
@@ -363,39 +370,175 @@ commands:
+      << : *build_atom_variant_shared_params
+      input_tag:
+        type: string
+    prev_image:
```
good catch; those were typos and should be removed.

dpipemazo commented on .circleci/config.yml:
```diff
@@ -585,10 +741,9 @@ workflows:
+          base_tag: *base_build_atom_opengl_cuda_tag
           production_image: nvidia/cuda:10.2-cudnn7-runtime-ubuntu18.04
-          build_args: --build-arg INSTALL_OPENGL=y --build-arg TEST_FROM=graphics
+          build_args: --build-arg INSTALL_OPENGL=y --build-arg
```
good catch

dpipemazo commented on Dockerfile-opengl:
```diff
@@ -0,0 +1,19 @@
+#
+################################################################################
+
+ARG BASE_IMAGE=elementaryrobotics/atom
```
we should likely remove the default here as it's confusing.
